sqlmesh/utils/lineage.py

```diff
@@ -0,0 +1,404 @@
+import typing as t
+from pathlib import Path
+
+from pydantic import Field
+
+from sqlmesh.core.dialect import normalize_model_name
+from sqlmesh.core.linter.helpers import (
+    TokenPositionDetails,
+)
+from sqlmesh.core.linter.rule import Range, Position
+from sqlmesh.core.model.definition import SqlModel, ExternalModel, PythonModel, SeedModel
+from sqlglot import exp
+from sqlglot.optimizer.scope import build_scope
+
+from sqlglot.optimizer.normalize_identifiers import normalize_identifiers
+from ruamel.yaml import YAML
+
+from sqlmesh.utils.pydantic import PydanticModel
+
+if t.TYPE_CHECKING:
+    from sqlmesh.core.context import Context
+    from sqlmesh.core.context import GenericContext
+
+
+class LSPModelReference(PydanticModel):
+    """A LSP reference to a model, excluding external models."""
+
+    type: t.Literal["model"] = "model"
+    path: Path
+    range: Range
+    markdown_description: t.Optional[str] = None
+
+
+class LSPExternalModelReference(PydanticModel):
+    """A LSP reference to an external model."""
+
+    type: t.Literal["external_model"] = "external_model"
+    range: Range
+    target_range: t.Optional[Range] = None
+    path: t.Optional[Path] = None
+    """The path of the external model, typically a YAML file, it is optional because
+    external models can be unregistered and so the path is not available."""
+
+    markdown_description: t.Optional[str] = None
+
+
+class LSPCteReference(PydanticModel):
```

Comment:
Provided that this is now in the common utils package, I wonder whether we should remove the `LSP` prefix from type names.